packages/core/src/plugins/default/markbind-plugin-tree.js

```diff
@@ -0,0 +1,176 @@
+    const lines = raw.split('\n').filter(line => line.trim() !== '');
+    const rootNode = new TreeNode('.', null, [], -1); // dummy root node
+    const prevParentStack = [rootNode];
+    let prevLevel = rootNode.level;
+    let prevParent = rootNode;
+    let prevNode = rootNode;
+    lines
+      .forEach((line) => {
+        const level = TreeNode.levelize(line);
+        const content = TreeNode.getContent(line);
+
+        if (level > prevLevel) {
+          prevParentStack.push(prevNode);
+          prevParent = prevNode;
+        } else if (level < prevLevel) {
+          for (let i = 0; i < prevLevel - level; i += 1) {
+            prevParentStack.pop();
+          }
+          prevParent = prevParentStack[prevParentStack.length - 1];
+        }
+
+        const newNode = new TreeNode(content, prevParent, [], level);
+        prevParent.children.push(newNode);
+        prevLevel = level;
+        prevNode = newNode;
+      });
+    return rootNode;
+  }
+
+  /**
+   * Traverses the tree and appends the tokens to the given array.
+   * @param {TreeNode} node - The node to traverse.
+   * @param {Array} treeTokens - The array to append the tokens to.
+   */
+  static traverse(currNode, result) {
+    if (!currNode.children) {
+      return;
+    }
+    if (currNode.parent === null) {
+      result.push(md.renderInline(`${currNode.content}\n`));
+    } else {
+      const tokens = [
+        '\n',
+        md.renderInline(currNode.content),
+        currNode.getPositionalToken(),
+      ];
+
+      // computes the strings appended to the content of the TreeNode
+      let curr = currNode.parent;
+      while (_.has(curr, 'parent.parent')) {
+        tokens.push(curr.isLastChild() ? TOKEN.space : TOKEN.connector);
+        curr = curr.parent;
+      }
+
+      result.push(tokens.reverse().join(''));
+    }
+    currNode.children.forEach((child) => {
+      TreeNode.traverse(child, result);
+    });
+  }
+
+  /**
+   * Returns the TreeNode as a string.
+   * This assumes that the node is a root node.
+   * @return {string}
+   */
+  toString() {
+    const treeTokens = [];
+    TreeNode.traverse(this, treeTokens);
+    return treeTokens.join('');
+  }
+
+  /**
+   * Returns the rendered tree.
+   * @param {string} raw - The raw text to parse.
+   * @return {string}
+   */
+  static visualize(raw) {
+    const dummyRootNode = TreeNode.parse(raw);
+    return dummyRootNode.children
+      .reduce((prev, curr) => {
+        curr.parent = null;
+        return prev + curr.toString();
+      }, '');
+  }
+}
+
+module.exports = {
+  tagConfig: {
+    tree: {
+      isSpecial: true,
+    },
+  },
+  getLinks: () => [`<link rel="stylesheet" href="${CSS_FILE_NAME}">`],
+  processNode: (pluginContext, node) => {
+    if (node.name !== 'tree') {
+      return;
+    }
+    node.name = 'div';
+    node.attribs.class = node.attribs.class ? `${node.attribs.class} tree` : 'tree';
+    node.children[0].data = TreeNode.visualize(node.children[0].data);
```

Comment:
DomParser nodes returned by htmlparser have various `type`s, with the `isSpecial` option specified the (only) child is treated as text (much like script/style tags).

You could use cheerio apis here instead to attach the visualized string as html, which solves the main (non-whitespace) issue with #1831.

Reply:
Debugger:
![Untitled](https://user-images.githubusercontent.com/3306138/158653548-586d7149-8668-4484-96a0-cf184b4e138c.png)

Reply:
> You could use cheerio apis here instead to attach the visualized string as html

Do you mean changing it to something like:
```js
  processNode: (pluginContext, node) => {
    if (node.name !== 'tree') {
      return;
    }
    cheerio(node).text(TreeNode.visualize(cheerio(node).text()));
    cheerio(node).addClass('tree');
    node.name = 'div';
  },
```
I tried this but doesn't seem to resolve the issue with tooltips (still has white space + does not show up on hover).

Reply:
yup, but this is the wrong method `.text(` (still attaching as text, not html)

Reply:
reping in case this was missed @tlylt.
though not urgent, we can put up an issue if no time to test it yet.

Reply:
> reping in case this was missed @tlylt. though not urgent, we can put up an issue if no time to test it yet.

Thanks for the reminder. I have tested and setting it via .html (iirc) does achieve the effect as you mentioned. This resolves the white-space issue but triggers are still quite buggy/ not functional. I have #1831 to keep track of this and was thinking to investigate along with the comment https://github.com/MarkBind/markbind/issues/1831#issuecomment-1068871935 before pushing a fix.